File: linked_list_cycle.java

//Problem Name: Linked List Cycle
//Difficulty: Easy


/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    
    // NOTE: This is the optimal solution
    // Solution using fast and slow pointers (Floyd's cycle detection algorithm)
    // Memory: O(1)
    // Runtime: O(n)
    public boolean hasCycle(ListNode head) {
        if (head == null) return false;
        
        ListNode slow = head;
        ListNode fast = head.next;
        
        // Use two pointers, fast and slow
        while(slow != fast){
            if(fast == null || fast.next == null){
                return false; // End of list is reached, no cycle is found
            }
            slow = slow.next;
            fast = fast.next.next;
        }
        return true;
    }
    
    
    // Solution using HashSet
    // Memory: O(n)
    // Runtime: O(n)
    public boolean hasCycle(ListNode head){
        HashSet<ListNode> seenNodes = new HashSet<ListNode>();
        if(head == null) return false;
        
        while(head != null){
            if(seenNodes.contains(head)){
                return true;
            }
            seenNodes.add(head);
            head = head.next;
        }
        return false;
    }
}

File: linked-lists/linkedListCycle.py
# Problem: Linked List Cycle


class ListNode:
    def __init__(self, val):
        self.val = val
        self.next = None


# Looks for cycle in by storing previously seen nodes in a set
# Runtime: O(n)
# Memory: O(n)
def hasCycleUsingSetImplementation(head):
    seenNodes = set()
    if head is None:
        return False
    while head is not None:
        if head in seenNodes:
            return True
        seenNodes.add(head)
        head = head.next
    return False


# Looks for cycle using fast and slow pointer (Sliding window)
# Runtime: O(n)
# Memory: O(1)
def hasCycleUsingSlidingWindow(head):

    if head is None:
        return False

    pointerA = head  # slow pointer
    pointerB = head.next  # fast pointer

    while pointerA.val != pointerB.val:
        # If you reach the end of the list without finding cycle, return false
        if pointerB is None or pointerB.next is None:
            return False
        pointerA = pointerA.next
        pointerB = pointerB.next.next
    return True


if __name__ == "__main__":
    print("Hello World!")
